src/transformers/models/beit3/modeling_beit3.py

```diff
@@ -0,0 +1,1197 @@
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, Beit3Encoder):
+            module.gradient_checkpointing = value
+
+    def reset_parameters(self):
+        self.fc1.reset_parameters()
+        self.fc2.reset_parameters()
+        if self.ffn_layernorm is not None:
+            self.ffn_layernorm.reset_parameters()
+
+
+class Beit3MultiwayFeedForwardNetwork(nn.Module):
+    def __init__(self, module):
+        super().__init__()
+        self.dim = 1
+        self.first = module
+        self.second = copy.deepcopy(module)
+        self.second.reset_parameters()
+
+    def forward(self, hidden_states, split_position=-1):
+        if split_position == -1:
+            return self.first(hidden_states)
+        if split_position == 0:
+            return self.second(hidden_states)
+        text_hidden, image_hidden = torch.split(
+            hidden_states,
+            [split_position, hidden_states.size(self.dim) - split_position],
+            dim=self.dim,
+        )
+        y1, y2 = self.first(text_hidden), self.second(image_hidden)
+        return torch.cat([y1, y2], dim=self.dim)
+
+
+class Beit3Linear(nn.Module):
+    def __init__(self, config):
+        super(Beit3Linear, self).__init__()
+        self.linear_1 = nn.Linear(config.embed_dim, config.embed_dim, bias=True)
+        self.linear_2 = nn.Linear(config.embed_dim, config.embed_dim, bias=True)
+
+    def forward(self, hidden_states, split_position=-1):
+        if split_position == -1:
+            return self.linear_1(hidden_states)
+        if split_position == 0:
+            return self.linear_2(hidden_states)
+        text_hidden, image_hidden = torch.split(
+            hidden_states,
+            [split_position, hidden_states.size(1) - split_position],
+            dim=1,
+        )
+        y1, y2 = self.linear_1(text_hidden), self.linear_2(image_hidden)
+        return torch.cat([y1, y2], dim=1)
+
+
+class Beit3LayerNorm(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.layernorm_1 = nn.LayerNorm(config.embed_dim, eps=config.layernorm_eps)
+        self.layernorm_2 = nn.LayerNorm(config.embed_dim, eps=config.layernorm_eps)
+
+    def forward(self, hidden_states, split_position=-1):
+        if split_position == -1:
+            return self.layernorm_1(hidden_states)
+
+        if split_position == 0:
+            return self.layernorm_2(hidden_states)
+
+        text_hidden, image_hidden = torch.split(
+            hidden_states,
+            [split_position, hidden_states.size(1) - split_position],
+            dim=1,
+        )
+        text_hidden = self.layernorm_1(text_hidden)
+        image_hidden = self.layernorm_2(image_hidden)
+        hidden_states = torch.cat([text_hidden, image_hidden], dim=1)
+        return hidden_states
+
+
+class Beit3Embedder(nn.Module):
+    def __init__(self, modules, dim=1):
```

Comment:
* As above - layers shouldn't take layers as arguments to instantiate themselves 
* All layers should take the config as a first argument and as few (or no) additional kwargs
* `dim` can be removed here